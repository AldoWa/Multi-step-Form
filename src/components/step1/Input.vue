<template>
  <div class="flex flex-col">
    <label 
      htmlFor="input"
      class="text-sm font-medium text-[#111827] mb-2"
    >
      {{ label }}
    </label>
    <input
      class="
        border-2 
        border-gray-300 
        rounded-[10px] 
        h-10
        p-2 
        focus:outline-none 
        focus:ring-2 
        focus:ring-[#FC6C4C] 
        focus:border-transparent
        placeholder-gray-400
        hover:cursor-pointer
      "
      :type="type" 
      name="input"
      id="input"
      :autocomplete="autocomplete"
      @input="emitValue"
      :value="modelValue"
      :placeholder="placeholder"
    >
  </div>
</template>

<script setup lang="ts">


interface InputProps {
  label: string
  modelValue: string
  autocomplete?: string;
  type: string;
  placeholder: string;
}

defineProps<InputProps>()
const emit = defineEmits(['update:modelValue'])

function emitValue(e: Event) {
  const target = e.target as HTMLInputElement
  emit('update:modelValue', target.value)
}

</script>
