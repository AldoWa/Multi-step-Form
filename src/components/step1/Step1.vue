<template>
  <div class="grid grid-cols-2 gap-6 mt-6">
    <Input 
      label="Full name" 
      v-model="personalInformationLocal.name"
      type="text"
      placeholder="Richard Rasmunse"
    />
    <Input 
      label="Email Address" 
      v-model="personalInformationLocal.email"
      type="email"
      autocomplete="email"
      placeholder="name@email.com"
    />
    <Input 
      label="Phone Number" 
      v-model="personalInformationLocal.phone"
      type="text"
      autocomplete="tel"
      placeholder="(79) 988129628"
    />
    <Input 
      label="Github Link" 
      v-model="personalInformationLocal.githubLink"
      type="text"
      autocomplete="url"
      placeholder="github.com/richardrasmunse"
    />
  </div>
</template>

<script setup lang="ts">
import {  ref, watch } from 'vue';

import Input from './Input.vue';
import { DataProvide } from '../../types/provide';

import useInject from '../../composables/inject';

type PersonalInformation = DataProvide['personalInformation']

const personalInformationLocal = ref<PersonalInformation>({
  name: '',
  email: '',
  phone: '',
  githubLink: '',
})

watch(personalInformationLocal, (newVal: PersonalInformation) => {
  changePersonalInformation(newVal)
}, { deep: true })

const { changePersonalInformation } = useInject()

</script>
