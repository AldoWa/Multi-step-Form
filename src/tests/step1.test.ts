import { describe, expect, it, vitest } from "vitest";

import { mount } from "@vue/test-utils";

import Input from "../components/step1/Input.vue";
import Step1 from "../components/step1/Step1.vue";

describe("Step 1", () => {
  describe('Input', () => {
    it('Should render the input with the correct type, placeholder and label', () => {
      const input = mount(Input, {
        props: {
          placeholder: 'Name Placeholder',
          label: 'Name Label',
          modelValue: '',
          type: 'text'
        }
      })

      const inputElement = input.get('[data-test="input-container"]')

      expect(inputElement.get('input').attributes('placeholder')).toBe('Name Placeholder');
      expect(inputElement.get('input').attributes('type')).toBe('text');

      expect(inputElement.get('label').text()).toBe('Name Label');
    })

    it('Should change the modelValue when the input changes', async () => {
      const wrapper = mount(Input, {
        props: {
          placeholder: 'Name Placeholder',
          label: 'Name Label',
          modelValue: 'initialText',
          type: 'text',
        }
      })

      await wrapper.setProps({
        'onUpdate:modelValue': (value) => wrapper.setProps({ modelValue: value }),
      })
      const input = wrapper.find('input')
      await input.setValue('New Value')

      expect(wrapper.props('modelValue')).toBe('New Value');
    })

    it('Should match with the snapshot', () => {
      const input = mount(Input, {
        props: {
          placeholder: 'Name Placeholder',
          label: 'Name Label',
          modelValue: '',
          type: 'text'
        }
      })

      expect(input.html()).toMatchSnapshot()
    })
  })

  describe('Step 1', () => {
    /**
     * It's not recommended to acess vm directly, but in this case, it's necessary to test the provider
     */
    it('Should add data to the provider', async () => {
      const wrapper = mount(Step1, {
        global: {
          provide: {
            data: {
              changePersonalInformation: vitest.fn((value) => {
                wrapper.vm.$.provides.data.personalInformation = value
              }),
              changeSkill: vitest.fn(),
              changePreference: vitest.fn(),
              personalInformation: {
                name: '',
                email: '',
                phone: '',
                githubLink: ''
              }
            }
          }
        }
      })

      const [name,email, phone, githubLink] = wrapper.findAll('input')

      await name.setValue('Name')
      await email.setValue('email@email.com')
      await phone.setValue('799999999')
      await githubLink.setValue('Link')

      expect(wrapper.vm.$.provides.data.personalInformation).toEqual({
        name: 'Name',
        email: 'email@email.com',
        phone: '(79) 9999-999',
        githubLink: 'Link'
      })
    })

    it('Should match with the snapshot', async () => {
      const wrapper = mount(Step1, {
        global: {
          provide: {
            data: {
              personalInformation: {
                name: '',
                email: '',
                phone: '',
                githubLink: ''
              }
            }
          }
        }
      })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})