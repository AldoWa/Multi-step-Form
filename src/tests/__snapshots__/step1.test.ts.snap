// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Step 1 > Input > Should match with the snapshot 1`] = `"<div class="flex flex-col" data-test="input-container"><label for="input" class="text-sm font-medium text-[#111827] mb-2">Name Label</label><input class="border-2 border-gray-300 rounded-[10px] h-10 p-2 focus:outline-none focus:ring-2 focus:ring-[#FC6C4C] focus:border-transparent placeholder-gray-400 hover:cursor-pointer" type="text" name="input" id="input" placeholder="Name Placeholder"></div>"`;

exports[`Step 1 > Step 1 > Should match with the snapshot 1`] = `
"<div class="grid grid-cols-2 gap-6">
  <div class="flex flex-col" data-test="input-container"><label for="input" class="text-sm font-medium text-[#111827] mb-2">Full name</label><input class="border-2 border-gray-300 rounded-[10px] h-10 p-2 focus:outline-none focus:ring-2 focus:ring-[#FC6C4C] focus:border-transparent placeholder-gray-400 hover:cursor-pointer" type="text" name="input" id="input" placeholder="Richard Rasmunse"></div>
  <div class="flex flex-col" data-test="input-container"><label for="input" class="text-sm font-medium text-[#111827] mb-2">Email Address</label><input class="border-2 border-gray-300 rounded-[10px] h-10 p-2 focus:outline-none focus:ring-2 focus:ring-[#FC6C4C] focus:border-transparent placeholder-gray-400 hover:cursor-pointer" type="email" name="input" id="input" autocomplete="email" placeholder="name@email.com"></div>
  <div class="flex flex-col" data-test="input-container"><label for="input" class="text-sm font-medium text-[#111827] mb-2">Phone Number</label><input class="border-2 border-gray-300 rounded-[10px] h-10 p-2 focus:outline-none focus:ring-2 focus:ring-[#FC6C4C] focus:border-transparent placeholder-gray-400 hover:cursor-pointer" type="tel" name="input" id="input" autocomplete="tel" placeholder="(79) 988129628"></div>
  <div class="flex flex-col" data-test="input-container"><label for="input" class="text-sm font-medium text-[#111827] mb-2">Github Link</label><input class="border-2 border-gray-300 rounded-[10px] h-10 p-2 focus:outline-none focus:ring-2 focus:ring-[#FC6C4C] focus:border-transparent placeholder-gray-400 hover:cursor-pointer" type="text" name="input" id="input" autocomplete="url" placeholder="github.com/richardrasmunse"></div>
</div>"
`;
